### ISSUE TITLE
cmake install on Mac OS 10.13.4 using intel@18.0.2 fails

### ISSUE THREAD

**Issue by JohnWGrove** [spack/spack#8175]
Summary
I tried to install pfunit using spack on my Mac Pro workstation using the intel 18.0.2 compiler. The installation failed when the dependency for cmake was attempted to be installed. The error was the the intel C++ compiler did not support C++11, which is false. By default this compiler enables the C++11 standard via the command line -std=c++11. The build script for cmake did not know enough about the Intel compiler to automatically add this argument. I probably could work around this problem, but it should have been handled transparently by me. In fact this issue brings up the general question of why spack thought it needed to install cmake in the first place since I already had a working version in my path via macports. I probably could also have added one via spack using clang@9.1.0. This seems to be a general issue with spack. When it needs a executable tool it just blindly installs one, without first checking whether it was really needed.  I assume this has something to do with the "build clean" philosophy spack uses since it probably clobbers the executable path as well as the rest of the environment.

Expected Result

Spack would do correctly what I requested and be smart enough to know about compiler options and executable tools.

### Actual Result

The installation failed for the reasons described above.

### Steps to reproduce the issue

I ran "spack install pfunit %intel@18.0.2" but the problem should be reproduced by
"spack install cmake %intel@18.0.2"

Incidentally, the installation use gnu@7.3.0 for cmake failed due to incompatibilities with the CoreGraphics framework on my mac and the installation for pfunit using clang also fails since the git site https://git.code.sf.net/p/pfunit/code was unavailable. Could this address be obsolete?

### Information on your system

Mac Pro late 2013 model running OS 10.13.4. I have the gnu compilers via spack and purchased licenses for the intel compilers, current version 18.0.2.

-----

We encourage you to try, as much as possible, to reduce your problem to the minimal example that still reproduces the issue. That would help us a lot in fixing it quickly and effectively!

If you want to ask a question about the tool (how to use it, what it can currently do, etc.), try the `#general` channel on our Slack first. We have a welcoming community and chances are you'll get your reply faster and without opening an issue.

Other than that, thanks for taking the time to contribute to Spack!

**Comment by jrood-nrel** [comment]
Newer versions of MacOS have objective-c code in the system headers, which causes cmake to fail building with GCC. Since Intel compilers typically use the GCC front end underneath, could this be the issue?

Also Spack thought it needed to build cmake itself, because the mechanism for that is to list cmake as an external in `packages.yaml` and then Spack will use it. This also usually needs to be done for pkg-config on MacOS as well. Pfunit might install after that.

**Comment by JohnWGrove** [comment]
Thanks Jon, the objective C business is likely related to the build problem. If you see this message before I dive through the documentation, could you tell me how to specify a package like cmake as external in packages.yaml. Thanks.

John W. Grove
CCS-2 MS D413, PO Box 1663
Los Alamos National Laboratory
Los Alamos, NM 87545
(505) 667-0723
jgrove@lanl.gov

**Comment by JohnWGrove** [comment]
Thanks a bunch Jon: This is very helpful.

John W. Grove
CCS-2 MS D413, PO Box 1663
Los Alamos National Laboratory
Los Alamos, NM 87545
(505) 667-0723
jgrove@lanl.gov

**Comment by JohnWGrove** [comment]
Jon: I have one other question. I have three compilers clang, gcc, and intel, when you have an external specification like

cmake:



paths:



cmake@3.11.1: /usr/local/Cellar/cmake/3.11.1


buildable: False

will this apply to all compilers or should I include separate paths for each version?

Specifically for me the question is whether something like

cmake:
    paths:
      cmake@3.11.1%gcc@7.3.0 arch=highsierra-x86_64: /opt/local
      cmake@3.11.1%clang@9.1.0-apple arch=highsierra-x86_64: /opt/local
      cmake@3.11.1%intel@18.0.2 arch=highsierra-x86_64: /opt/local

or will

cmake:
    paths:
      cmake@3.11.1 arch=highsierra-x86_64: /opt/local

handle all the cases.
Also, I suspect the arch specification is really not needed since spack seems to be able to figure this out itself.




John W. Grove
CCS-2 MS D413, PO Box 1663
Los Alamos National Laboratory
Los Alamos, NM 87545
(505) 667-0723
jgrove@lanl.gov

**Comment by adamjstewart** [comment]
@mathstuf

**Comment by jrood-nrel** [comment]
@JohnWGrove Yes, being less specific about the spec for each external package should cover more ground. So for something like CMake, this is fine to use for all compilers. I have also been leaving the arch out as it seems to be unnecessary. I haven't tried this yet, but now I'm even wondering if taking the version of the external package specification is possible, or if it needs it to register the package version in the database. That way any new CMake version that showed up in Spack would be expected to never cause a rebuild of CMake. Anyway I think your first specification for CMake should be good.

**Comment by adamjstewart** [comment]
@jrood-nrel I think version is the only thing that's actually required aside from the package name. It's also not a good idea to leave it out. A lot of CMake packages require a newer version of CMake than is typically installed on the OS.

**Comment by jrood-nrel** [comment]
@adamjstewart Cool, thanks. I'll keep the version in. Just never thought about that until now.

**Comment by JohnWGrove** [comment]
Thanks very much for your help. I give all of this a try.

John W. Grove
CCS-2 MS D413, PO Box 1663
Los Alamos National Laboratory
Los Alamos, NM 87545
(505) 667-0723
jgrove@lanl.gov

**Comment by citibeth** [comment]
Are you using a version of Spack that incorporates #7994?

**Comment by JohnWGrove** [comment]
Hi Elizabeth: I don’t recall the version, but I will upgrade to current version and try again.

John W. Grove
CCS-2 MS D413, PO Box 1663
Los Alamos National Laboratory
Los Alamos, NM 87545
(505) 667-0723
jgrove@lanl.gov

**Comment by citibeth** [comment]
> Hi Elizabeth: I don’t recall the version, but I will upgrade to current version and try again.

It looks like your problem is with CMake not pFUnit, so I doubt it would make a difference.  But I agree, using the latest pFUnit recipe is a Good Thing.

In general, my experience with Mac is, a LOT of things require Clang to build, due to the Objective-C stuff.  Python, for example.  And CMake too, I suppose.  The "usual" solution to that problem is to use the built-in compiler for C and C++, and a different compiler for Fortran; a "mixed toolchain".  There has been a lot of success doing this with Clange (C/C++) + GFortran, I expect that mixing Clang + Intel Fortran would also work.  Depending on your situation, that might work better than setting certain packages as external.

One problem with the external CMake solution here is, what if your system CMake isn't new enough for some packages?  That can be solved by using Spack to build the latest CMake (with Clang), and then setting up that installed version as "external."

If the concretizer were more clever, it would allow build dependencies to be built with a different compiler than the thing being built.  This would be the most flexible solution to your problem.

**Comment by citibeth** [comment]
John...

1.5 years ago, I ran into a lot of these same problems.  The most expedient
solution for me was to have IT install Linux in place of MacOS on my big
desktop Mac.  I've been very happy ever since.

I do a little bit of Spack on my Mac laptop; but no heavy lifting.
Certainly nothing that would require the Intel compiler.

My recommendation would be to seriously consider switching to Linux, if you
can.  Macs present problems, and Intel compilers present problems.  Both at
once, and you're dealing with problems squared.

-- Elizabeth
